include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -1090,23 +1475,82 @@ struct __brick_reduce_idx
-               _OutValues& __out_values) const
+    __scalar_path_impl(_IsFull, const _ItemId __idx, const _ReduceIdx& __segment_starts, const _Values& __values,
+                       _OutValues& __out_values) const
     {
         using __value_type = decltype(__segment_starts[__idx]);
         __value_type __segment_end =
             (__idx == __segment_starts.size() - 1) ? __value_type(__n) : __segment_starts[__idx + 1];
         __out_values[__idx] = reduce(__segment_starts[__idx], __segment_end, __values);
     }
+    template <typename _IsFull, typename _ItemId, typename _ReduceIdx, typename _Values, typename _OutValues>
+    void
+    operator()(_IsFull __is_full, const _ItemId __idx, const _ReduceIdx& __segment_starts, const _Values& __values,
+               _OutValues& __out_values) const
+    {
+        __scalar_path_impl(__is_full, __idx, __segment_starts, __values, __out_values);
+    }
 
   private:
     _BinaryOperator __binary_op;
     _Size __n;
 };
 
+// std::swap_ranges is unique in that both sets of provided ranges will be modified. Due to this,
+// we define a separate functor from __walk2_vectors_or_scalars with a customized vectorization path.
+template <typename _ExecutionPolicy, typename _F, typename _Range1, typename _Range2>
+struct __brick_swap : public walk_vector_or_scalar_base<_Range1, _Range2>
+{
+  private:
+    using __base_t = walk_vector_or_scalar_base<_Range1, _Range2>;
+    _F __f;
+    std::size_t __n;
+
+  public:
+    __brick_swap(_F __f, std::size_t __n) : __f(__f), __n(__n) {}
+
+    template <typename _IsFull>
+    void
+    __vector_path_impl(_IsFull __is_full, const std::size_t __idx, _Range1 __rng1, _Range2 __rng2) const
+    {
+        using _ValueType1 = oneapi::dpl::__internal::__value_t<_Range1>;
+        using _ValueType2 = oneapi::dpl::__internal::__value_t<_Range2>;
+        _ValueType1 __rng1_vector[__base_t::__preferred_vector_size];
+        _ValueType1 __rng2_vector[__base_t::__preferred_vector_size];
+        // 1. Load inputs into vectors
+        oneapi::dpl::__par_backend_hetero::__vector_load<__base_t::__preferred_vector_size>{__n}(
+            __is_full, __idx, oneapi::dpl::__par_backend_hetero::__scalar_load_op{}, __rng1, __rng1_vector);
+        oneapi::dpl::__par_backend_hetero::__vector_load<__base_t::__preferred_vector_size>{__n}(
+            __is_full, __idx, oneapi::dpl::__par_backend_hetero::__scalar_load_op{}, __rng2, __rng2_vector);
+        // 2. Swap the two ranges
+        oneapi::dpl::__par_backend_hetero::__vector_store<__base_t::__preferred_vector_size>{__n}(
+            __is_full, __idx, oneapi::dpl::__par_backend_hetero::__scalar_store_transform_op<_F>{__f}, __rng2_vector,
+            __rng1);
+        oneapi::dpl::__par_backend_hetero::__vector_store<__base_t::__preferred_vector_size>{__n}(
+            __is_full, __idx, oneapi::dpl::__par_backend_hetero::__scalar_store_transform_op<_F>{__f}, __rng1_vector,
```

Comment:
So, we are applying this swap function twice from the vector to the output range, as transform operator for both store operations.  
Couldn't we just swap the vectors with each other, then store each to the output range, 

I don't think we can skip the "swap" operation, and just write from each "other" loaded vector, but that is probably fastest, and we could check on that.

Reply:
Actually, even better: 
Can we just load `__rng1` into a vector, swap the vector with `__rng2`, then store the now "already swapped" vector to `__rng1`?

Reply:
> Actually, even better: Can we just load `__rng1` into a vector, swap the vector with `__rng2`, then store the now "already swapped" vector to `__rng1`?

Benchmarked both suggestions and saw similar performance. I chose this one since it is cleaner and has lower register pressure.